src/workerd/api/sockets.c++

```diff
@@ -66,7 +75,11 @@ jsg::Ref<Socket> setupSocket(
   // Initialise the readable/writable streams with the readable/writable sides of an AsyncIoStream.
   auto sysStreams = newSystemMultiStream(refcountedConnection->addWrappedRef(), ioContext);
   auto readable = jsg::alloc<ReadableStream>(ioContext, kj::mv(sysStreams.readable));
-  readable->initEofResolverPair(js);
+  auto allowHalfOpen = getAllowHalfOpen(options);
+  kj::Maybe<jsg::Promise<void>> eofPromise;
+  if (!allowHalfOpen) {
+    eofPromise = readable->onEof(js);
+  }
```

Comment:
To Kenton's point about delivering this sooner... I think it would make sense to not use a promise here actually. Just use a callback that can be invoked synchronously when `signalEof()` is called and have the callback close the writable immediately rather than deferring a promise task to do so, keeping in mind, however, that the writable should be flushed before it is actually closed.

Reply:
I could see an argument for a callback _if_ it called all the way back to JavaScript so that the application could synchronously set a flag to tell itself not to write anymore.

However, that would be a different API and I'm not sure it's an API I particularly like.

If we don't change the API, so we're still returning a promise to the app, then I don't think it matters whether we use a promise or a callback for the internal part.

Reply:
Yeah, I think this is fine. It's likely possible for us to optimize this later if necessary without it being a breaking change.